src/gretel_trainer/relational/json.py

```diff
@@ -336,5 +335,19 @@ def _generate_commands(
     return (_add_single_table, add_foreign_key)
 
 
+def get_json_columns(df: pd.DataFrame) -> list[str]:
+    column_previews = {
+        col: df[col].dropna().head(5)
+        for col in df.columns
+        if df.dtypes[col] == "object"
+    }
+
+    return [
+        col
+        for col, series in column_previews.items()
+        if series.apply(is_dict).all() or series.apply(is_list).all()
```

Comment:
This is a really interesting call-out.

First, current context: the `if/elif` works currently even with a mix of dicts and lists because if we enter the `if dict_cols` branch, we transform the dict values (lifting them to columns) and add the resulting df to the `nested_dfs` list; then on the next iteration of the function we shouldn't find any dicts, but we will find the (unaltered, original) list columns and deal with them then. What's interesting, though, is that this means we're parsing the list columns the first time through, but then dropping the result on the floor and picking it up again the next time.

I tried switching the `elif` to an `if` and a bunch of stuff blows up—I think trying to combine the two transformations in one iteration through the function is too complicated.

Instead, I'm pushing up a change that defers the check for any list_columns so that if we find any dict columns first, we just deal with them immediately. We'll still have a few iterations of the function that check both, but at least it'll be fewer?

Reply:
I'm happy to talk through the logic here if easier. The code currently works with dicts within lists and lists within dicts, but the top-level of each column must be either all list, or all dict. I don't think anything is dropped on the floor - I could be wrong though. A table might be processed multiple times because it contains a list inside a dict inside a list inside a dict, etc. so this flattening each of these is handled in recursive steps (where the next run sees the output of the previous run, i.e. the same table with one less level of nesting).

Reply:
In other words the code expects JSON with some sort of reasonable schema. Optional/non-existent params/columns are OK, but the same param/column (including both top-level columns and invented columns) cannot be a list sometimes, and a dict other times. Handling the latter is quite complex and may require making some assumptions which may not always be correct, and I'm hoping that most production datasets are much better behaved than that, so I left it out for now (left it out here means that the JSON special handling is not applied to these columns, so they are treated as categorical).

Reply:
Looking at the code again, this line might be unnecessarily computed `list_cols = [col for col in cols_to_scan if df[col].dropna().apply(is_list).all()]` if dict_cols is not empty, so depending on how expensive it is, it might make sense to defer computing it until after we know that `dict_cols` is empty (change the `elif` to `else: list_cols=...` then `if list_cols:`.

Reply:
@sami-t exactly, that's the change I made in [623e15b](https://github.com/gretelai/trainer/pull/112/commits/623e15b8165cd3460466f77dc33f1baa78aa0d36)